src/System Application/App/File System/src/Scenario/FileScenarioImpl.Codeunit.al

```diff
@@ -0,0 +1,329 @@
+
+            if FileAccountScenarios.FindSet() then
+                repeat
+                    // Add entry for every scenario that is assigned to the current file account
+                    AddEntry(Result, FileAccountScenarios.EntryType::Scenario, FileAccountScenarios.Scenario, FileAccountScenarios."Account Id", FileAccountScenarios.Connector, FileAccountScenarios."Display Name", false, Position);
+                until FileAccountScenarios.Next() = 0;
+        until FileAccounts.Next() = 0;
+
+        // Order by position to show accurate results
+        Result.SetCurrentKey(Position);
+    end;
+
+    local procedure GetFileScenariosForAccount(FileAccount: Record "File Account"; var FileAccountScenarios: Record "File Account Scenario")
+    var
+        FileScenarios: Record "File Scenario";
+        ValidFileScenarios: DotNet Hashtable;
+        IsScenarioValid: Boolean;
+        Scenario: Integer;
+    begin
+        FileAccountScenarios.Reset();
+        FileAccountScenarios.DeleteAll();
+
+        // Get all file scenarios assigned to the file account
+        FileScenarios.SetRange("Account Id", FileAccount."Account Id");
+        FileScenarios.SetRange(Connector, FileAccount.Connector);
+
+        if not FileScenarios.FindSet() then
+            exit;
+
+        // Find all valid scenarios. Invalid scenario may occur if the extension that added them was removed.
+        ValidFileScenarios := ValidFileScenarios.Hashtable();
+        foreach Scenario in Enum::"File Scenario".Ordinals() do
+            ValidFileScenarios.Add(Scenario, Scenario);
+
+        // Convert File Scenario-s to File Account Scenario-s so they can be sorted by "Display Name"
+        repeat
+            IsScenarioValid := ValidFileScenarios.Contains(FileScenarios.Scenario.AsInteger());
+
+            // Add entry for every scenario that exists and uses the file account. Skip the default scenario.
+            if (FileScenarios.Scenario <> Enum::"File Scenario"::Default) and IsScenarioValid then begin
+                FileAccountScenarios.Scenario := FileScenarios.Scenario.AsInteger();
+                FileAccountScenarios."Account Id" := FileScenarios."Account Id";
+                FileAccountScenarios.Connector := FileScenarios.Connector;
+                FileAccountScenarios."Display Name" := Format(FileScenarios.Scenario);
+
+                FileAccountScenarios.Insert();
+            end;
+        until FileScenarios.Next() = 0;
+
+        FileAccountScenarios.SetCurrentKey("Display Name"); // sort scenarios by "Display Name"
+    end;
+
+    local procedure AddEntry(var Result: Record "File Account Scenario"; EntryType: Option; Scenario: Integer; AccountId: Guid; Connector: Enum "File System Connector"; DisplayName: Text[2048]; Default: Boolean; var Position: Integer)
+    begin
+        // Add entry to the result while maintaining the position so that the tree represents the data correctly
```

Comment:
Clean() since its has var, to make sure it always inserts correct values

Reply:
I don't know what the Clean() function is. Can you describe it in more detail?

Reply:
I think @StefanSosic is referring to clear method: https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/system/system-clear-joker-method

If you add this a the beginning of the procedure then the the result record we be cleared.
```al
Clear(Result);
```

Reply:
Yeah, exactly, sorry for confusion 😕 
@IceOnly  Clear method
Thanks @pri-kise

Reply:
Now I understand.
I copied the whole scenario and account setup from the email module:
https://github.com/microsoft/BCApps/blob/33d32690f347e273c4e952926e28464dd2da19af/src/System%20Application/App/Email/src/Scenario/EmailScenarioImpl.Codeunit.al#L165

Normaly an init instead of clear should make sence. But I think it was done that way on purpose, otherwise I can't explain the comment in the function.

Reply:
Init clears only PKs. I get it that lot of code was copied, but you need to understand what are procedures which you copied doing and how. Copying because it's done like that on another place is not g

Reply:
That's not true, init reset every field except the PK, but you can say include PK:
https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-init-method

I know of no case where a clear on a record make sense.

> Copying because it's done like that on another place is not good practice...

That is exactly that I told my trainess and developers.
But the Email module is used by thousand of installations and we have test code for it.
The functions a a realy specific comment:
> // Add entry to the result while maintaining the position so that the tree represents the data correctly
So I thought to myself, this will solve a very specific problem that we are not aware of. The code works as expected for me.

Solving the same problem with different implantation in the system apps seems to be a bad habit for me.

Reply:
I added an Init().

Reply:
That's what I wanted to say. Non PKs. 
It could be some edge case, not covered by tests.

True, but that would be than next step, to open another issue to adjust other places to use "better" approach.

I will check and do testing, maybe you could do and try to break current approach

Reply:
> I added an Init().

Sounds good to me